homeassistant/components/unifi/switch.py

```diff
@@ -301,430 +488,136 @@ async def options_updated(self) -> None:
-        """Call when controller connection state change."""
-        self.async_signalling_callback(ItemEvent.ADDED, self._obj_id)
+        obj = description.object_fn(self.controller.api, obj_id)
+        self._attr_is_on = description.is_on_fn(controller.api, obj)
+        self._attr_name = description.name_fn(obj)
 
     async def async_turn_on(self, **kwargs: Any) -> None:
         """Turn on connectivity for client."""
-        await self.controller.api.request(
-            ClientBlockRequest.create(self._obj_id, False)
+        await self.entity_description.control_fn(
+            self.controller.api, self._obj_id, True
         )
 
     async def async_turn_off(self, **kwargs: Any) -> None:
         """Turn off connectivity for client."""
-        await self.controller.api.request(ClientBlockRequest.create(self._obj_id, True))
-
-    @property
-    def icon(self) -> str:
-        """Return the icon to use in the frontend."""
-        if not self.is_on:
-            return "mdi:network-off"
-        return "mdi:network"
-
-    async def options_updated(self) -> None:
-        """Config entry options are updated, remove entity if option is disabled."""
-        if self._obj_id not in self.controller.option_block_clients:
-            await self.remove_item({self._obj_id})
-
-    async def remove_item(self, keys: set) -> None:
-        """Remove entity if key is part of set."""
-        if self._obj_id not in keys or self._removed:
-            return
-        self._removed = True
-        if self.registry_entry:
-            er.async_get(self.hass).async_remove(self.entity_id)
-        else:
-            await self.async_remove(force_remove=True)
-
-
-class UnifiDPIRestrictionSwitch(SwitchEntity):
-    """Representation of a DPI restriction group."""
-
-    _attr_entity_category = EntityCategory.CONFIG
-
-    def __init__(self, obj_id: str, controller: UniFiController) -> None:
-        """Set up dpi switch."""
-        controller.entities[DOMAIN][DPI_SWITCH].add(obj_id)
-        self._obj_id = obj_id
-        self.controller = controller
-
-        dpi_group = controller.api.dpi_groups[obj_id]
-        self._known_app_ids = dpi_group.dpiapp_ids
-
-        self._attr_available = controller.available
-        self._attr_is_on = self.calculate_enabled()
-        self._attr_name = dpi_group.name
-        self._attr_unique_id = dpi_group.id
-        self._attr_device_info = DeviceInfo(
-            entry_type=DeviceEntryType.SERVICE,
-            identifiers={(DOMAIN, f"unifi_controller_{obj_id}")},
-            manufacturer=ATTR_MANUFACTURER,
-            model="UniFi Network",
-            name="UniFi Network",
+        await self.entity_description.control_fn(
+            self.controller.api, self._obj_id, False
         )
 
     async def async_added_to_hass(self) -> None:
         """Register callback to known apps."""
+        handler = self.entity_description.api_handler_fn(self.controller.api)
```

Comment:
This might be a bit easier to read if you did
```suggestion
        description = self.entity_description
        handler = description.api_handler_fn(self.controller.api)
```

...